Begin adding abstractions for implementing WebTransport

achristensen07 opened the pull request:
#### 328209aba28e0d37d98f3561c05c20c2906f7780
<pre>
Begin adding abstractions for implementing WebTransport
<a href="https://bugs.webkit.org/show_bug.cgi?id=261007">https://bugs.webkit.org/show_bug.cgi?id=261007</a>
rdar://114800387

Reviewed by Youenn Fablet.

This isn&apos;t complete.  It&apos;s notably missing the part where it actually hooks up to something that sends and receives over the network.
But local experiments have shown that this is enough infrastructure to get datagrams and streams sending and receiving, and we can polish from there.
This is a good starting point.

Several abstractions are necessary.  This introduces them and begins hooking them up to each other, but they are not completely hooked up yet.

We need a way to get a WebKit::WebTransportSession from WebCore.  I used SocketProvider for this.  It&apos;s a very similar job to providing WebSocket channels.
It will be renamed accordingly.

We need a way to get from C++ into the various ReadableStreams when datagrams or streams arrive.  I made WebTransportBidirectionalStreamSource
and DatagramSource and WebTransportReceiveStreamSource for this.

We need a way to get datagrams from the JS WritableStream to C++.  I made DatagramSink for this.

We need a way to get from the web process to the network process and back.  I made WebTransportSession and NetworkTransportSession for this,
and I made them communicate with each other and have linked lifetimes.  They own WebTransportStream and NetworkTransportStream objects which
I did not make MessageReceivers and MessageSenders because their identifiers need to be ProcessQualified because sometimes they are made in the
web process and sometimes their creation is initiated from the network process.  So like WebFrame/WebFrameProxy (which also needs to have an identifier
that is ProcessQualified) to send a message to these objects you send a message to their owner that then routes the message to the owned object.

The initial values for some of the member variables of WebTransportDatagramDuplexStream are stated by the specification to be implementation defined.
Chrome and Firefox both have an initial value of 1024 for m_outgoingMaxDatagramSize, so I decided to go with the same value.
Chrome and Firefox both have an initial value of 1 for m_incomingDatagramsHighWaterMark, so I did the same.
Firefox has an initial value of 5 for m_outgoingDatagramsHighWaterMark and Chrome seems to effectively have an initial value of 1, so I went with 1.
These values can be revisited if necessary.

It builds, it shows a little bit of progress in the tests, and it seems like things are mostly in the right place, so this is a good checkpoint to make a PR from.
Nothing is actually hooked up to the network yet, but that&apos;ll be the next step.

* LayoutTests/imported/w3c/web-platform-tests/webtransport/connect.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webtransport/constructor.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webtransport/csp-fail.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webtransport/csp-pass.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webtransport/datagram-cancel-crash.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webtransport/echo-large-bidirectional-streams.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webtransport/in-removed-iframe.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webtransport/streams-close.https.any-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webtransport/constructor.https.any-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webtransport/DatagramSink.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportCloseInfo.h.
(WebCore::DatagramSink::write):
(WebCore::DatagramSink::close):
(WebCore::DatagramSink::error):
* Source/WebCore/Modules/webtransport/DatagramSink.h: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.h.
(WebCore::DatagramSink::create):
* Source/WebCore/Modules/webtransport/DatagramSource.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportCloseInfo.h.
(WebCore::DatagramSource::setActive):
(WebCore::DatagramSource::setInactive):
(WebCore::DatagramSource::doStart):
(WebCore::DatagramSource::doPull):
(WebCore::DatagramSource::doCancel):
* Source/WebCore/Modules/webtransport/DatagramSource.h: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.h.
(WebCore::DatagramSource::create):
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::create):
(WebCore::WebTransport::initializeOverHTTP):
(WebCore::createPromiseAndWrapper):
(WebCore::WebTransport::WebTransport):
(WebCore::WebTransport::activeDOMObjectName const):
(WebCore::WebTransport::virtualHasPendingActivity const):
(WebCore::WebTransport::receiveDatagram):
(WebCore::WebTransport::receiveIncomingUnidirectionalStream):
(WebCore::WebTransport::receiveBidirectionalStream):
(WebCore::WebTransport::getStats):
(WebCore::WebTransport::ready):
(WebCore::WebTransport::reliability):
(WebCore::WebTransport::congestionControl):
(WebCore::WebTransport::closed):
(WebCore::WebTransport::draining):
(WebCore::WebTransport::datagrams):
(WebCore::WebTransport::incomingBidirectionalStreams):
(WebCore::WebTransport::incomingUnidirectionalStreams):
(WebCore::createRejectingPromise): Deleted.
* Source/WebCore/Modules/webtransport/WebTransport.h:
* Source/WebCore/Modules/webtransport/WebTransport.idl:
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamSource.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.cpp.
(WebCore::WebTransportBidirectionalStreamSource::setActive):
(WebCore::WebTransportBidirectionalStreamSource::setInactive):
(WebCore::WebTransportBidirectionalStreamSource::doStart):
(WebCore::WebTransportBidirectionalStreamSource::doPull):
(WebCore::WebTransportBidirectionalStreamSource::doCancel):
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamSource.h: Copied from Source/WebCore/Modules/webtransport/WebTransportError.h.
(WebCore::WebTransportBidirectionalStreamSource::create):
* Source/WebCore/Modules/webtransport/WebTransportCloseInfo.h:
* Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.cpp:
(WebCore::WebTransportDatagramDuplexStream::create):
(WebCore::WebTransportDatagramDuplexStream::WebTransportDatagramDuplexStream):
(WebCore::WebTransportDatagramDuplexStream::readable):
(WebCore::WebTransportDatagramDuplexStream::writable):
(WebCore::WebTransportDatagramDuplexStream::maxDatagramSize):
(WebCore::WebTransportDatagramDuplexStream::incomingMaxAge):
(WebCore::WebTransportDatagramDuplexStream::outgoingMaxAge):
(WebCore::WebTransportDatagramDuplexStream::incomingHighWaterMark):
(WebCore::WebTransportDatagramDuplexStream::outgoingHighWaterMark):
(WebCore::WebTransportDatagramDuplexStream::setIncomingMaxAge):
(WebCore::WebTransportDatagramDuplexStream::setOutgoingMaxAge):
(WebCore::WebTransportDatagramDuplexStream::setIncomingHighWaterMark):
(WebCore::WebTransportDatagramDuplexStream::setOutgoingHighWaterMark):
* Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.h:
* Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.idl:
* Source/WebCore/Modules/webtransport/WebTransportError.h:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.cpp.
(WebCore::WebTransportReceiveStreamSource::setActive):
(WebCore::WebTransportReceiveStreamSource::setInactive):
(WebCore::WebTransportReceiveStreamSource::doStart):
(WebCore::WebTransportReceiveStreamSource::doPull):
(WebCore::WebTransportReceiveStreamSource::doCancel):
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.h: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.h.
(WebCore::WebTransportReceiveStreamSource::create):
* Source/WebCore/Modules/webtransport/WebTransportSendStream.cpp:
(WebCore::WebTransportSendStream::sendOrder):
(WebCore::WebTransportSendStream::setSendOrder):
* Source/WebCore/Modules/webtransport/WebTransportSendStream.h:
* Source/WebCore/Modules/webtransport/WebTransportSession.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportCloseInfo.h.
(WebCore::WebTransportSession::attachClient):
* Source/WebCore/Modules/webtransport/WebTransportSession.h: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.h.
* Source/WebCore/Modules/webtransport/WebTransportSessionClient.h: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.h.
(WebCore::WebTransportSessionClient::~WebTransportSessionClient):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/page/SocketProvider.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::didReceiveMessage):
(WebKit::NetworkConnectionToWebProcess::initializeWebTransportSession):
(WebKit::NetworkConnectionToWebProcess::destroyWebTransportSession):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkTransportSession.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.cpp.
(WebKit::NetworkTransportSession::NetworkTransportSession):
(WebKit::NetworkTransportSession::sendDatagram):
(WebKit::NetworkTransportSession::streamSendBytes):
(WebKit::NetworkTransportSession::streamSendStopSending):
(WebKit::NetworkTransportSession::streamReset):
(WebKit::NetworkTransportSession::createOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::createBidirectionalStream):
* Source/WebKit/NetworkProcess/NetworkTransportSession.h: Copied from Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.h.
* Source/WebKit/NetworkProcess/NetworkTransportSession.messages.in: Added.
* Source/WebKit/NetworkProcess/NetworkTransportStream.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.h.
(WebKit::NetworkTransportStream::sendBytes):
(WebKit::NetworkTransportStream::sendStopSending):
(WebKit::NetworkTransportStream::reset):
* Source/WebKit/NetworkProcess/NetworkTransportStream.h: Copied from Source/WebCore/Modules/webtransport/WebTransportCloseInfo.h.
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::didReceiveMessage):
* Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
(WebKit::WebSocketProvider::initializeWebTransportSession):
* Source/WebKit/WebProcess/Network/WebSocketProvider.h:
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp: Added.
(WebKit::WebTransportSession::initialize):
(WebKit::WebTransportSession::WebTransportSession):
(WebKit::WebTransportSession::~WebTransportSession):
(WebKit::WebTransportSession::messageSenderConnection const):
(WebKit::WebTransportSession::messageSenderDestinationID const):
(WebKit::WebTransportSession::receiveDatagram):
(WebKit::WebTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::WebTransportSession::receiveBidirectionalStream):
(WebKit::WebTransportSession::streamReceiveBytes):
(WebKit::WebTransportSession::sendDatagram):
(WebKit::WebTransportSession::createOutgoingUnidirectionalStream):
(WebKit::WebTransportSession::createBidirectionalStream):
* Source/WebKit/WebProcess/Network/WebTransportSession.h: Added.
* Source/WebKit/WebProcess/Network/WebTransportSession.messages.in: Added.
* Source/WebKit/WebProcess/Network/WebTransportStream.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportCloseInfo.h.
(WebKit::WebTransportStream::receiveBytes):
* Source/WebKit/WebProcess/Network/WebTransportStream.h: Copied from Source/WebCore/Modules/webtransport/WebTransportCloseInfo.h.
* Source/WebKit/WebProcess/Network/WorkerWebTransportSession.cpp: Added.
(WebKit::WorkerWebTransportSession::initialize):
(WebKit::WorkerWebTransportSession::WorkerWebTransportSession):
(WebKit::WorkerWebTransportSession::~WorkerWebTransportSession):
(WebKit::WorkerWebTransportSession::receiveDatagram):
(WebKit::WorkerWebTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::WorkerWebTransportSession::receiveBidirectionalStream):
(WebKit::WorkerWebTransportSession::sendDatagram):
(WebKit::WorkerWebTransportSession::createOutgoingUnidirectionalStream):
(WebKit::WorkerWebTransportSession::createBidirectionalStream):
* Source/WebKit/WebProcess/Network/WorkerWebTransportSession.h: Copied from Source/WebCore/Modules/webtransport/WebTransportError.h.
* Source/WebKit/WebProcess/WebCoreSupport/WebBadgeClient.cpp:
(WebKit::WebBadgeClient::setAppBadge):
(WebKit::WebBadgeClient::setClientBadge):
* Source/WebKit/WebProcess/WebPage/MomentumEventDispatcher.cpp:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::webTransportSession):
(WebKit::WebProcess::addWebTransportSession):
(WebKit::WebProcess::removeWebTransportSession):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKitLegacy/WebCoreSupport/LegacySocketProvider.cpp:
(LegacySocketProvider::initializeWebTransportSession):
* Source/WebKitLegacy/WebCoreSupport/LegacySocketProvider.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/328209aba28e0d37d98f3561c05c20c2906f7780

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19712 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20135 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20752 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21607 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18427 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19948 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23395 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20278 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19998 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19932 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19935 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17148 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22461 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/19896 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17114 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17940 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24223 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18182 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18114 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22207 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18709 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15854 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17863 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22217 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2412 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18534 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/webtransport/DatagramSink.cpp:
```diff
@@ -0,0 +1,71 @@
+        return completionHandler();
+    RefPtr session = transport->session();
+    if (!session)
+        return completionHandler();
```
Shouldn't we throw in case transport is dead?
We might also want to log in case of !session, maybe reject as well?
Maybe this is too early as the spec asks to create a queue. Maybe we should add a FIXME.

youennf commented on Source/WebCore/Modules/webtransport/DatagramSink.cpp:
```diff
@@ -0,0 +1,71 @@
+
+void DatagramSink::attachTo(WebTransport& transport)
+{
+    m_transport = transport;
```
ASSERT(! m_transport) I would guess as we do not want to switch

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -26,89 +26,317 @@
+    auto bidirectionalStreamSource = WebTransportBidirectionalStreamSource::create();
+    auto incomingBidirectionalStreams = ReadableStream::create(domGlobalObject, bidirectionalStreamSource.copyRef());
+    if (incomingBidirectionalStreams.hasException()) {
+        ASSERT_NOT_REACHED();
```
Aren't we sometimes getting exceptions in case of workers being terminated while creating a stream?
Probably ditto below

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -26,89 +26,317 @@
+
+void WebTransport::getStats(Ref<DeferredPromise>&& promise)
+{
+    promise->reject(nullptr);
```
Exception NotSupported maybe.

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -26,89 +26,317 @@
-void WebTransport::close(std::optional<WebTransportCloseInfo>&&)
+static CString trimToValidUTF8Length1024(CString&& string)
 {
+    if (string.length() > 1024)
```
Shouldn't we do a fast path for below 1024?

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -26,89 +26,317 @@
+    m_sendStreams = { }; // FIXME: Error each send stream.
+    // FIXME: The use of cancel and the use of NetworkError aren't quite correct in this function.
+    for (auto& stream : std::exchange(m_receiveStreams, { }))
+        stream->cancel(Exception { NetworkError });
```
We probably want to error through the stream controller., I guess the FIXMEs cover this.

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -26,89 +26,317 @@
+    for (auto& stream : std::exchange(m_receiveStreams, { }))
+        stream->cancel(Exception { NetworkError });
+
+    if (closeInfo) {
```
```
if (!closeInfo) {
    ....
    return;
}
```

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -26,89 +26,317 @@
+        // FIXME: Resolve with closeInfo.
+        m_closed.second->resolve();
+        m_incomingBidirectionalStreams->cancel(Exception { NetworkError });
+        m_incomingUnidirectionalStreams->cancel(Exception { NetworkError });
```
Probably close here instead of cancel.

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.cpp:
```diff
@@ -26,89 +26,317 @@
+        return promise->reject(InvalidStateError);
+    m_session->createBidirectionalStream([promise = WTFMove(promise), context = WeakPtr { context }] (std::optional<BidirectionalStreamConstructionParameters>&& parameters) {
+        if (!parameters)
+            return promise->reject(nullptr);
```
It is strange to reject will nullptr, shouldn't we throw with an Exception or bail out?

youennf commented on Source/WebCore/Modules/webtransport/WebTransport.h:
```diff
@@ -34,32 +39,86 @@ class JSGlobalObject;
+    void cleanup(Ref<DOMException>&&, std::optional<WebTransportCloseInfo>&&);
+
+    const char* activeDOMObjectName() const final;
+    bool virtualHasPendingActivity() const final;
```
Should we add suspend() that would close the connection?

youennf commented on Source/WebCore/Modules/webtransport/WebTransportReceiveStream.cpp:
```diff
@@ -27,12 +27,33 @@
-    promise->reject(nullptr);
+    m_statsProvider->getStats([promise = WTFMove(promise)] (std::optional<WebTransportReceiveStreamStats>&& stats) mutable {
+        if (!stats)
+            return promise->reject(nullptr);
```
exception ?

youennf commented on Source/WebCore/Modules/webtransport/WebTransportReceiveStream.cpp:
```diff
@@ -27,12 +27,33 @@
+        if (!stats)
+            return promise->reject(nullptr);
+        // FIXME: Resolve promise with stats.
+        return promise->reject(nullptr);
```
not supported error

youennf commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -286,7 +288,13 @@ void NetworkConnectionToWebProcess::didReceiveMessage(IPC::Connection& connectio
-    
+
+    if (decoder.messageReceiverName() == Messages::NetworkTransportSession::messageReceiverName()) {
+        if (auto* networkTransportSession = m_networkTransportSessions.get(WebTransportSessionIdentifier(decoder.destinationID())))
```
Should we from the start try to do networking in a workqueue (or soon)?

youennf commented on Source/WebCore/page/SocketProvider.h:
```diff
@@ -31,12 +31,15 @@
 class WEBCORE_EXPORT SocketProvider : public ThreadSafeRefCounted<SocketProvider> {
 public:
     virtual RefPtr<ThreadableWebSocketChannel> createWebSocketChannel(Document&, WebSocketChannelClient&) = 0;
+    virtual void initializeWebTransportSession(ScriptExecutionContext&, const URL&, CompletionHandler<void(RefPtr<WebTransportSession>&&)>&&) = 0;
```
It is a bit strange to reuse WebSocket provider for WebTransport.
Shouldn't we use a separate WebTransportProvider?
Or maybe rename WebSocketProvider in WebNetworkProvider?

youennf commented on Source/WebKit/WebProcess/Network/WorkerWebTransportReceiveStreamStatsProvider.cpp:
```diff
@@ -0,0 +1,65 @@
+void WorkerWebTransportReceiveStreamStatsProvider::getStats(CompletionHandler<void(std::optional<WebCore::WebTransportReceiveStreamStats>&&)>&& completionHandler)
+{
+    ASSERT(!RunLoop::isMain());
+    RunLoop::main().dispatch([provider = m_provider, identifier = m_scriptExecutionContextIdentifier, completionHandler = WTFMove(completionHandler)] () mutable {
```
Given we might want to do networking from a worker thread without going to main thread, I wonder whether getStats will not follow the same path.
How about removing this code for now?

youennf commented on Source/WebKit/WebProcess/Network/WorkerWebTransportSendStreamSink.cpp:
```diff
@@ -0,0 +1,69 @@
+
+    auto bytes = WTF::switchOn(arrayBufferOrView, [&](auto& arrayBufferOrView) {
+        return Vector<uint8_t> { std::span<const uint8_t> { static_cast<const uint8_t*>(arrayBufferOrView->data()), arrayBufferOrView->byteLength() } };
+    });
```
A lot of this code will be the same in window and worker.
It might be best to reuse the same code as much as possible.
Ideally, the implementation (worker or not) would just be to make some IPC from either worker or main threads.
This will reduce copies and be faster as well, which is important for WebTransport realtime use cases.

youennf reviewed (approved):
Main point to me is that we should, from the start, try to go from worker thread to networking process work queue (and conversely) since WebTransport might be used for perf intensive media streaming.

LGTM provided we remove the worker specific bits that induce the model to go to main thread.

achristensen07 committed:
Begin adding abstractions for implementing WebTransport
https://bugs.webkit.org/show_bug.cgi?id=261007
rdar://114800387

Reviewed by Youenn Fablet.

This isn't complete.  It's notably missing the part where it actually hooks up to something that sends and receives over the network.
But local experiments have shown that this is enough infrastructure to get datagrams and streams sending and receiving, and we can polish from there.
This is a good starting point.

Several abstractions are necessary.  This introduces them and begins hooking them up to each other, but they are not completely hooked up yet.

We need a way to get a WebKit::WebTransportSession from WebCore.  I used SocketProvider for this.  It's a very similar job to providing WebSocket channels.
It will be renamed accordingly.

We need a way to get from C++ into the various ReadableStreams when datagrams or streams arrive.  I made WebTransportBidirectionalStreamSource
and DatagramSource and WebTransportReceiveStreamSource for this.

We need a way to get datagrams from the JS WritableStream to C++.  I made DatagramSink for this.

We need a way to get from the web process to the network process and back.  I made WebTransportSession and NetworkTransportSession for this,
and I made them communicate with each other and have linked lifetimes.  They own WebTransportStream and NetworkTransportStream objects which
I did not make MessageReceivers and MessageSenders because their identifiers need to be ProcessQualified because sometimes they are made in the
web process and sometimes their creation is initiated from the network process.  So like WebFrame/WebFrameProxy (which also needs to have an identifier
that is ProcessQualified) to send a message to these objects you send a message to their owner that then routes the message to the owned object.

The initial values for some of the member variables of WebTransportDatagramDuplexStream are stated by the specification to be implementation defined.
Chrome and Firefox both have an initial value of 1024 for m_outgoingMaxDatagramSize, so I decided to go with the same value.
Chrome and Firefox both have an initial value of 1 for m_incomingDatagramsHighWaterMark, so I did the same.
Firefox has an initial value of 5 for m_outgoingDatagramsHighWaterMark and Chrome seems to effectively have an initial value of 1, so I went with 1.
These values can be revisited if necessary.

It builds, it shows a little bit of progress in the tests, and it seems like things are mostly in the right place, so this is a good checkpoint to make a PR from.
Nothing is actually hooked up to the network yet, but that'll be the next step.

* LayoutTests/imported/w3c/web-platform-tests/webtransport/connect.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webtransport/constructor.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webtransport/csp-fail.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webtransport/csp-pass.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webtransport/datagram-cancel-crash.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webtransport/echo-large-bidirectional-streams.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webtransport/in-removed-iframe.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webtransport/streams-close.https.any-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webtransport/constructor.https.any-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webtransport/DatagramSink.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportCloseInfo.h.
(WebCore::DatagramSink::write):
(WebCore::DatagramSink::close):
(WebCore::DatagramSink::error):
* Source/WebCore/Modules/webtransport/DatagramSink.h: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.h.
(WebCore::DatagramSink::create):
* Source/WebCore/Modules/webtransport/DatagramSource.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportCloseInfo.h.
(WebCore::DatagramSource::setActive):
(WebCore::DatagramSource::setInactive):
(WebCore::DatagramSource::doStart):
(WebCore::DatagramSource::doPull):
(WebCore::DatagramSource::doCancel):
* Source/WebCore/Modules/webtransport/DatagramSource.h: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.h.
(WebCore::DatagramSource::create):
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::create):
(WebCore::WebTransport::initializeOverHTTP):
(WebCore::createPromiseAndWrapper):
(WebCore::WebTransport::WebTransport):
(WebCore::WebTransport::activeDOMObjectName const):
(WebCore::WebTransport::virtualHasPendingActivity const):
(WebCore::WebTransport::receiveDatagram):
(WebCore::WebTransport::receiveIncomingUnidirectionalStream):
(WebCore::WebTransport::receiveBidirectionalStream):
(WebCore::WebTransport::getStats):
(WebCore::WebTransport::ready):
(WebCore::WebTransport::reliability):
(WebCore::WebTransport::congestionControl):
(WebCore::WebTransport::closed):
(WebCore::WebTransport::draining):
(WebCore::WebTransport::datagrams):
(WebCore::WebTransport::incomingBidirectionalStreams):
(WebCore::WebTransport::incomingUnidirectionalStreams):
(WebCore::createRejectingPromise): Deleted.
* Source/WebCore/Modules/webtransport/WebTransport.h:
* Source/WebCore/Modules/webtransport/WebTransport.idl:
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamSource.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.cpp.
(WebCore::WebTransportBidirectionalStreamSource::setActive):
(WebCore::WebTransportBidirectionalStreamSource::setInactive):
(WebCore::WebTransportBidirectionalStreamSource::doStart):
(WebCore::WebTransportBidirectionalStreamSource::doPull):
(WebCore::WebTransportBidirectionalStreamSource::doCancel):
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamSource.h: Copied from Source/WebCore/Modules/webtransport/WebTransportError.h.
(WebCore::WebTransportBidirectionalStreamSource::create):
* Source/WebCore/Modules/webtransport/WebTransportCloseInfo.h:
* Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.cpp:
(WebCore::WebTransportDatagramDuplexStream::create):
(WebCore::WebTransportDatagramDuplexStream::WebTransportDatagramDuplexStream):
(WebCore::WebTransportDatagramDuplexStream::readable):
(WebCore::WebTransportDatagramDuplexStream::writable):
(WebCore::WebTransportDatagramDuplexStream::maxDatagramSize):
(WebCore::WebTransportDatagramDuplexStream::incomingMaxAge):
(WebCore::WebTransportDatagramDuplexStream::outgoingMaxAge):
(WebCore::WebTransportDatagramDuplexStream::incomingHighWaterMark):
(WebCore::WebTransportDatagramDuplexStream::outgoingHighWaterMark):
(WebCore::WebTransportDatagramDuplexStream::setIncomingMaxAge):
(WebCore::WebTransportDatagramDuplexStream::setOutgoingMaxAge):
(WebCore::WebTransportDatagramDuplexStream::setIncomingHighWaterMark):
(WebCore::WebTransportDatagramDuplexStream::setOutgoingHighWaterMark):
* Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.h:
* Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.idl:
* Source/WebCore/Modules/webtransport/WebTransportError.h:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.cpp.
(WebCore::WebTransportReceiveStreamSource::setActive):
(WebCore::WebTransportReceiveStreamSource::setInactive):
(WebCore::WebTransportReceiveStreamSource::doStart):
(WebCore::WebTransportReceiveStreamSource::doPull):
(WebCore::WebTransportReceiveStreamSource::doCancel):
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.h: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.h.
(WebCore::WebTransportReceiveStreamSource::create):
* Source/WebCore/Modules/webtransport/WebTransportSendStream.cpp:
(WebCore::WebTransportSendStream::sendOrder):
(WebCore::WebTransportSendStream::setSendOrder):
* Source/WebCore/Modules/webtransport/WebTransportSendStream.h:
* Source/WebCore/Modules/webtransport/WebTransportSession.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportCloseInfo.h.
(WebCore::WebTransportSession::attachClient):
* Source/WebCore/Modules/webtransport/WebTransportSession.h: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.h.
* Source/WebCore/Modules/webtransport/WebTransportSessionClient.h: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.h.
(WebCore::WebTransportSessionClient::~WebTransportSessionClient):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/page/SocketProvider.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::didReceiveMessage):
(WebKit::NetworkConnectionToWebProcess::initializeWebTransportSession):
(WebKit::NetworkConnectionToWebProcess::destroyWebTransportSession):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkTransportSession.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.cpp.
(WebKit::NetworkTransportSession::NetworkTransportSession):
(WebKit::NetworkTransportSession::sendDatagram):
(WebKit::NetworkTransportSession::streamSendBytes):
(WebKit::NetworkTransportSession::streamSendStopSending):
(WebKit::NetworkTransportSession::streamReset):
(WebKit::NetworkTransportSession::createOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::createBidirectionalStream):
* Source/WebKit/NetworkProcess/NetworkTransportSession.h: Copied from Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.h.
* Source/WebKit/NetworkProcess/NetworkTransportSession.messages.in: Added.
* Source/WebKit/NetworkProcess/NetworkTransportStream.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportSendStream.h.
(WebKit::NetworkTransportStream::sendBytes):
(WebKit::NetworkTransportStream::sendStopSending):
(WebKit::NetworkTransportStream::reset):
* Source/WebKit/NetworkProcess/NetworkTransportStream.h: Copied from Source/WebCore/Modules/webtransport/WebTransportCloseInfo.h.
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::didReceiveMessage):
* Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
(WebKit::WebSocketProvider::initializeWebTransportSession):
* Source/WebKit/WebProcess/Network/WebSocketProvider.h:
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp: Added.
(WebKit::WebTransportSession::initialize):
(WebKit::WebTransportSession::WebTransportSession):
(WebKit::WebTransportSession::~WebTransportSession):
(WebKit::WebTransportSession::messageSenderConnection const):
(WebKit::WebTransportSession::messageSenderDestinationID const):
(WebKit::WebTransportSession::receiveDatagram):
(WebKit::WebTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::WebTransportSession::receiveBidirectionalStream):
(WebKit::WebTransportSession::streamReceiveBytes):
(WebKit::WebTransportSession::sendDatagram):
(WebKit::WebTransportSession::createOutgoingUnidirectionalStream):
(WebKit::WebTransportSession::createBidirectionalStream):
* Source/WebKit/WebProcess/Network/WebTransportSession.h: Added.
* Source/WebKit/WebProcess/Network/WebTransportSession.messages.in: Added.
* Source/WebKit/WebProcess/Network/WebTransportStream.cpp: Copied from Source/WebCore/Modules/webtransport/WebTransportCloseInfo.h.
(WebKit::WebTransportStream::receiveBytes):
* Source/WebKit/WebProcess/Network/WebTransportStream.h: Copied from Source/WebCore/Modules/webtransport/WebTransportCloseInfo.h.
* Source/WebKit/WebProcess/Network/WorkerWebTransportSession.cpp: Added.
(WebKit::WorkerWebTransportSession::initialize):
(WebKit::WorkerWebTransportSession::WorkerWebTransportSession):
(WebKit::WorkerWebTransportSession::~WorkerWebTransportSession):
(WebKit::WorkerWebTransportSession::receiveDatagram):
(WebKit::WorkerWebTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::WorkerWebTransportSession::receiveBidirectionalStream):
(WebKit::WorkerWebTransportSession::sendDatagram):
(WebKit::WorkerWebTransportSession::createOutgoingUnidirectionalStream):
(WebKit::WorkerWebTransportSession::createBidirectionalStream):
* Source/WebKit/WebProcess/Network/WorkerWebTransportSession.h: Copied from Source/WebCore/Modules/webtransport/WebTransportError.h.
* Source/WebKit/WebProcess/WebCoreSupport/WebBadgeClient.cpp:
(WebKit::WebBadgeClient::setAppBadge):
(WebKit::WebBadgeClient::setClientBadge):
* Source/WebKit/WebProcess/WebPage/MomentumEventDispatcher.cpp:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::webTransportSession):
(WebKit::WebProcess::addWebTransportSession):
(WebKit::WebProcess::removeWebTransportSession):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKitLegacy/WebCoreSupport/LegacySocketProvider.cpp:
(LegacySocketProvider::initializeWebTransportSession):
* Source/WebKitLegacy/WebCoreSupport/LegacySocketProvider.h:

Canonical link: https://commits.webkit.org/268453@main
